olive/passes/onnx/bnb_quantization.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+import re
+from typing import Any, Dict, List
+
+import onnx
+from onnxruntime import __version__ as OrtVersion
+from packaging import version
+
+from olive.hardware import AcceleratorSpec
+from olive.model import ONNXModel
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.pass_config import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class OnnxBnb4Quantization(Pass):
+    """Quantize MatMul nodes in ONNX model using 4bit FP4/NF4 quantization."""
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "quant_type": PassConfigParam(
+                type_=str,
+                description="The quantization type. Only 'fp4' and 'nf4' are supported.",
+            ),
+            "quantized_modules": PassConfigParam(
+                type_=List[str],
+                description=(
+                    "The list of modules to quantize. Node names will be matched as '.*[./]{module}[./]MatMul$'. If not"
+                    " specified, all MatMul nodes will be quantized except those specified in nodes_to_exclude."
+                ),
+            ),
+            "nodes_to_exclude": PassConfigParam(
+                type_=list,
+                default_value=None,
+                description="List of node names to exclude from quantization.",
+            ),
+        }
+        config.update(get_external_data_config())
+        return config
+
+    def _run_for_config(
+        self, model: ONNXModel, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModel:
+        assert version.parse(OrtVersion) >= version.parse(
+            "1.16.2"
+        ), "MatMulBnb4Quantizer is only supported in onnxruntime >= 1.16.2"
+
+        from onnxruntime.quantization.matmul_bnb4_quantizer import MatMulBnb4Quantizer
+
+        output_model_path = ONNXModel.resolve_path(output_model_path)
+
+        quant_type = config["quant_type"]
+        quantized_modules = config["quantized_modules"]
+        if model.model_attributes:
+            quantized_modules = quantized_modules or model.model_attributes.get("quantized_modules")
+
+            # extract quant_type from model_attributes if not specified in config
+            if not quant_type:
+                quant_config = model.model_attributes.get("quantization_config")
+
+                if quant_config.get("load_in_8bit"):
+                    raise ValueError("load_in_8bit is not supported. Only 4-bit quantization is supported.")
+
+                quant_type = quant_config.get("bnb_4bit_quant_type")
+                if not quant_type:
+                    raise ValueError("quant_type is required.")
```

Comment:
bnb_4bit_quant_type?

Reply:
`qaunt_type` is the actual pass config parameter name. `bnb_4bit_quant_type` is only used to extract the value from quantization_config. 
If the pass cannot find it in the model attributes, then they should provide it as a pass config parameter.